subsys/net/lib/sockets/getaddrinfo.c

```diff
@@ -346,6 +353,86 @@ static inline int z_vrfy_z_zsock_getaddrinfo_internal(const char *host,
 
 #endif /* defined(CONFIG_DNS_RESOLVER) */
 
+#if defined(CONFIG_NET_IPV6)
+static int try_resolve_literal_ipv6(const char *host, const char *service,
+		const struct zsock_addrinfo *hints,
+		struct zsock_addrinfo *res)
+{
+	int family = AF_UNSPEC;
+	long port = 0;
+	int ret;
+	char *host_copy;
+	int socktype = SOCK_STREAM;
+	int protocol = IPPROTO_TCP;
+
+	if (hints) {
+		family = hints->ai_family;
+		if (hints->ai_socktype == SOCK_DGRAM) {
+			socktype = SOCK_DGRAM;
+			protocol = IPPROTO_UDP;
+		}
+	}
+
+	if (service) {
+		port = strtol(service, NULL, 10);
+		if (port < 1 || port > 65535) {
+			return DNS_EAI_NONAME;
+		}
+	}
+
+	if (host == NULL) {
+		/* Cannot parse literal IPv6 address if not given. */
+		return DNS_EAI_NONAME;
+	}
+
+	/* Check for IPv6 literal address. Start with a quick heuristic check,
+	 * of first char of the address, then do long validating inet_pton()
+	 * call if needed.
+	 */
+	if (family == AF_INET) {
+		/* Cannot parse IPv4 */
+		return DNS_EAI_NONAME;
+	}
+
+	if (host[0] != '[' || host[strlen(host) - 1] != ']') {
```

Comment:
There already exits `net_ipaddr_parse()` that provides similar functionality (parsing IPv[4|6] string). Could we utilize that functionality here so we do not need to call any malloc etc.

Reply:
I have not found this function. Thanks for suggestion!

Reply:
It looks this module could be refactored and literal IPv4 addresses could be checked alongside IPv6, instead of within DNS procedure. `net_ipaddr_parse()` parses IPv6 and IPv4 at once.

I would leave it as `TODO`, because I don't have any applications to verify if literal IPv4 and DNS resolvers work as expected. Someone working with IPv4 could do that later.